compiler-rt/lib/scudo/standalone/tests/combined_test.cpp

```diff
@@ -1077,6 +1077,133 @@ struct TestQuarantineSizeClassConfig {
   static const scudo::uptr SizeDelta = 0;
 };
 
+struct TestZeroOnDeallocConfig {
+  static const bool MaySupportMemoryTagging = false;
+  static const bool EnableZeroOnDealloc = true;
+
+  template <class A> using TSDRegistryT = scudo::TSDRegistrySharedT<A, 1U, 1U>;
+  struct Primary {
+    // Tiny allocator, its Primary only serves chunks of four sizes.
+    using SizeClassMap =
+        scudo::FixedSizeClassMap<TestQuarantineSizeClassConfig>;
+    static const scudo::uptr RegionSizeLog = DeathRegionSizeLog;
+    static const scudo::s32 MinReleaseToOsIntervalMs = INT32_MIN;
+    static const scudo::s32 MaxReleaseToOsIntervalMs = INT32_MAX;
+    typedef scudo::uptr CompactPtrT;
+    static const scudo::uptr CompactPtrScale = 0;
+    static const bool EnableRandomOffset = true;
+    static const scudo::uptr MapSizeIncrement = 1UL << 18;
+    static const scudo::uptr GroupSizeLog = 18;
+  };
+
+  template <typename Config>
+  using PrimaryT = scudo::SizeClassAllocator64<Config>;
+
+  struct Secondary {
+    template <typename Config>
+    using CacheT = scudo::MapAllocatorNoCache<Config>;
+  };
+
+  template <typename Config> using SecondaryT = scudo::MapAllocator<Config>;
+};
+
+struct TestNoZeroOnDeallocConfig : public TestZeroOnDeallocConfig {
+  static const bool EnableZeroOnDealloc = false;
+};
+
+TEST(ScudoCombinedTest, ZeroOnDeallocDisabled) {
+  // When option `EnableZeroOnDealloc` is false the memeory is not cleared on
+  // deallocation.
+  using AllocatorT = TestAllocator<TestNoZeroOnDeallocConfig>;
+  auto Allocator = std::unique_ptr<AllocatorT>(new AllocatorT());
+
+  constexpr scudo::uptr AllocatedSize = 1024;
+
+  void *P = Allocator->allocate(AllocatedSize, Origin);
+  ASSERT_NE(P, nullptr);
+  memset(P, 'B', AllocatedSize);
+
+  Allocator->deallocate(P, Origin);
+
+  for (scudo::uptr I = 1; I < AllocatedSize; ++I) {
+    // Verifies the memory was left unchanged.
+    ASSERT_EQ(static_cast<char *>(P)[I], 'B') << "at position " << I;
+  }
+}
```

Comment:
We shouldn't suppose what the behavior is when zero-on-dealloc is disabled. For example, if the deallocation triggers a page releasing than all the contents are still cleared.